client/app/scripts/selectors/chartSelectors.js

```diff
@@ -0,0 +1,124 @@
+import debug from 'debug';
+import { createSelector, createSelectorCreator, defaultMemoize } from 'reselect';
+import { Map as makeMap, is, Set } from 'immutable';
+
+import { getAdjacentNodes } from '../utils/topology-utils';
+
+
+const log = debug('scope:selectors');
+
+
+//
+// "immutable" createSelector
+//
+const createDeepEqualSelector = createSelectorCreator(
+  defaultMemoize,
+  is
+);
+
+
+const identity = v => v;
+
+
+const allNodesSelector = state => state.get('nodes');
+
+
+export const nodesSelector = createSelector(
+  allNodesSelector,
+  (allNodes) => allNodes.filter(node => !node.get('filtered'))
+);
+
+
+//
+// This is like an === cache...
+//
+// - getAdjacentNodes is run on every state change and can generate a new immutable object each
+// time:
+//   - v1 = getAdjacentNodes(a)
+//   - v2 = getAdjacentNodes(a)
+//   - v1 !== v2
+//   - is(v1, v2) === true
+//
+// - createDeepEqualSelector will wrap those calls with a: is(v1, v2) ? v1 : v2
+//   - Thus you can compare consecutive calls to adjacentNodesSelector(state) with === (which is
+//     what redux is doing with connect()
+//
+// Note: this feels like the wrong way to be using reselect...
```

Comment:
Yes, you should be able to rewrite `getAdjacentNodes` so it takes `nodes` (filtered) and `origin` as inputs. Then those two become the selectors, and `getAdjacentNodes` will be fully memoizable.